lib/apiservers/engine/.godeps_rules

```diff
@@ -0,0 +1,33 @@
+# By default, packages in lib should depend only on packages under pkg or vendor
+^pkg/*
+^vendor/*
+
+# lib/apiservers/engine subpackages can depend on other subpackages
+^lib/apiservers/engine/*
+
+# lib/apiservers/engine acts as a client of the portlayer API
+^lib/apiservers/portlayer/client/*
+^lib/apiservers/portlayer/models/*
+
+# lib/apiservers/engine also depends on portlayer utilities and events
+^lib/portlayer/util/*
+^lib/portlayer/event/events/*
+
+# lib/apiservers/engine also depends on other lib packages
+^lib/archive/*
+^lib/config/*
+^lib/constants/*
+^lib/guest/*
```

Comment:
I don't see this being directly included in the `lib/apiservers/engine` package when searching in my codebase.
I _assume_ if it was transient it would be in a different section.

This does bring up a valid extension/enhancement - if a package is listed as a dependency but is _not_ included I think we should also flag it. Otherwise we have a difficult time keeping the list to the minimum necessary set.

Reply:
`lib/guest` is used indirectly by `lib/apiservers/engine/backends/middleware`. It looks like I added it to this file before I noticed the weird transitive dependencies pulled in by `lib/dns`, and then overlooked it when separating those into a separate section. I've moved it to the correct location.

> This does bring up a valid extension/enhancement - if a package is listed as a dependency but is not included I think we should also flag it. Otherwise we have a difficult time keeping the list to the minimum necessary set.

I considered this, but my naive implementation of it caused problems. For example, `lib/.godeps_rules ` says that by default anything in `lib` can depend on anything in `pkg`. For many packages, this is sufficient. However, it's not always necessary — and not depending on something under `pkg` isn't a problem; the rule is still appropriate.

A better implementation might verify that each rule is used by at least one of the packages that uses the rules file (but not necessarily all of them), but that gets quite a bit more complicated.

Reply:
I found a way to handle detection of unused rules that isn't _too_ ugly. I've pushed it to this PR for discussion.

However, it raises some interesting withi questions:
1. How should generated code be handled? My original PR included rules to cover those packages so developers could run this check in a dirty environment. However, in a clean environment (e.g., CI), those files don't exist when checks are run and the pattern appears to be unused.
2. How do you communicate future intent? I think it's useful to be able to say "it would be okay for this to depend on `pkg/*`" even if the package does not currently have such a dependency.
3. When a package has an erroneous dependency, it's convenient to be able to just copy the parent directory's rules and tweak accordingly. With detection of unused rules, it becomes necessary to "prune" as well.

Perhaps some additional syntax could be introduced to communicate "it's okay if this pattern is unused", but the check for unused patterns is already more complicated than the enforcement logic.

Perhaps unused dependencies should be valid if they're being "inherited" (i.e., if they're present in the parent directory's rules file), but maybe actually supporting inheritance would be a better investment.

I do see accumulation of extraneous dependency information as a potential maintainability issue… I'm just not sure how best to handle it.

Reply:
> How should generated code be handled? My original PR included rules to cover those packages so developers could run this check in a dirty environment. However, in a clean environment (e.g., CI), those files don't exist when checks are run and the pattern appears to be unused.

I've addressed this be expressing a dependency on the swagger generation targets. It's not ideal (`make check` now takes longer as a result), but seems like a reasonable compromise.

> How do you communicate future intent? I think it's useful to be able to say "it would be okay for this to depend on pkg/*" even if the package does not currently have such a dependency.

I've addressed this by simply commenting out the relevant line; it can simply be uncommented when needed.

> When a package has an erroneous dependency, it's convenient to be able to just copy the parent directory's rules and tweak accordingly. With detection of unused rules, it becomes necessary to "prune" as well.

I decided that this was more related to the above question than I originally thought; both are about expressing what's _okay_ to depend on (even if no such dependence exists right now). I've also addressed this by commenting out unused, but acceptable, rules.